### DISCUSSION TITLE
Drag-and-drop reordering with dnd-kit

### DISCUSSION THREAD

**Discussion by aaronadamsCA** [edmundhung/conform#426] [Show and tell]
Here's my first take on a drag-and-drop reorder component using the Conform reorder intent:

https://stackblitz.com/edit/conform-dnd-kit-reorder?file=app%2Froutes%2F_index%2Froute.tsx

![image](https://github.com/edmundhung/conform/assets/1505561/d2255f52-dbc8-4da8-b690-0313db760540)

This provides a `SortableFieldList` component using `@dnd-kit/sortable` and the Conform `FormContext` API.

The biggest challenge was addressing the fact that you need to use `getFieldList().map((metadata) => {})` when rendering the list of items to ensure you can set `key={metadata.key}` on each one. Without this, the reorder intent doesn't work. But the metadata contains only the ID of the sortable component, without the rest of the data needed for rendering.

My simple solution was this component interface:

```ts
interface SortableFieldListProps {
  name: FieldName<UniqueIdentifier[]>;
  nodes: Map<UniqueIdentifier, ReactNode>;
}
```

- In the route, you create a `Map` associating each ID with a React node, then pass only its `keys()` to Conform's `defaultValue`.
- In the component, you loop through `getFieldList()` and `get()` the React node from the `Map`, which is very fast.

Since this uses a Conform intent and `parseWithZod`, everything should still work even without a client-side `onValidate` function! The visual transitions would just need a bit of work.

Let me know what you think and if you have any ideas for improvements.

**Comment by f-miyazaki** [comment]
Hi, I just wanted to say thanks for your helpful implementation example. It was exactly what I needed and helped me solve a problem I was working on. Really appreciate it!
